hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferContainer.java

```diff
@@ -417,22 +440,55 @@ private long addInternal(Data dataItem) {
     }
 
     /**
-     * Deserializes the dataItem if the ring buffer configured {@code inMemoryFormat} is set to {@code OBJECT}.
+     * Converts the {@code item} into the ringbuffer {@link InMemoryFormat} or keeps
+     * it unchanged if the supplied argument is already in the ringbuffer format.
      *
-     * @param dataItem the item in binary format
+     * @param item the item
      * @return the binary or deserialized format, depending on the {@link RingbufferContainer#inMemoryFormat}
-     * @throws com.hazelcast.nio.serialization.HazelcastSerializationException if the ring buffer is configured to keep items
-     *                                                                         in object format and the item could not be
-     *                                                                         deserialized
+     * @throws HazelcastSerializationException if the ring buffer is configured to keep items
+     *                                         in object format and the item could not be deserialized
+     */
+    private Object convertToRingbufferFormat(Object item) {
+        return inMemoryFormat == OBJECT
+                ? serializationService.toObject(item)
+                : serializationService.toData(item);
+    }
+
+    /**
+     * Convert the supplied argument into serialized format.
+     *
+     * @throws HazelcastSerializationException when serialization fails.
+     */
+    private Data convertToData(Object item) {
+        return serializationService.toData(item);
+    }
+
+    /**
+     * Convert the supplied argument into serialized format.
+     *
+     * @throws HazelcastSerializationException when serialization fails.
      */
-    private Object getRingbufferFormat(Data dataItem) {
-        return inMemoryFormat == OBJECT ? serializationService.toObject(dataItem) : dataItem;
+    private Data[] convertToData(T[] items) {
+        if (items == null || items.length == 0) {
+            return new Data[0];
+        }
+        if (items[0] instanceof Data) {
+            return (Data[]) items;
+        }
+        final Data[] ret = new Data[items.length];
+        for (int i = 0; i < items.length; i++) {
+            ret[i] = convertToData(items[i]);
+        }
+        return ret;
     }
 
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
+        assert !out.getVersion().isUnknown();
         boolean ttlEnabled = expirationPolicy != null;
-        out.writeUTF(name);
+        if (out.getVersion().isLessThan(Versions.V3_9)) {
+            out.writeUTF(namespace.getObjectName());
```

Comment:
The `name` in is not necessary since the `ReplicationOperation` has the `name` as well and provides it in the `RingbufferContainer#init` method. 
Hence 3.9 writes it only when the cluster version is lower than 3.9 and ignores it on read.